sylt/src/formatter.rs

```diff
@@ -0,0 +1,504 @@
+            write!(dest, "(")?;
+            if exprs.is_empty() {
+                write!(dest, ",")?;
+            } else {
+                write_comma_separated!(dest, indent, write_expression, exprs);
+            }
+            write!(dest, ")")?;
+        }
+        ExpressionKind::List(exprs) => {
+            write!(dest, "[")?;
+            write_comma_separated!(dest, indent, write_expression, exprs);
+            write!(dest, "]")?;
+        }
+        ExpressionKind::Set(exprs) => {
+            write!(dest, "{{")?;
+            write_comma_separated!(dest, indent, write_expression, exprs);
+            write!(dest, "}}")?;
+        }
+        ExpressionKind::Dict(exprs) => {
+            write!(dest, "{{")?;
+            let mut first = true;
+            let mut exprs = exprs.iter();
+            while let Some(expr) = exprs.next() {
+                if !first {
+                    write!(dest, ", ")?;
+                }
+                first = false;
+                write_expression(dest, indent, expr)?;
+                write!(dest, ": ")?;
+                write_expression(dest, indent, exprs.next().unwrap())?;
+            }
+            write!(dest, "}}")?;
+        }
+        ExpressionKind::Float(f) => write!(dest, "{}", f)?,
+        ExpressionKind::Int(i) => write!(dest, "{}", i)?,
+        ExpressionKind::Str(s) => write!(dest, "\"{}\"", s)?,
+        ExpressionKind::Bool(b) => write!(dest, "{}", if *b { "true" } else { "false" })?,
+        ExpressionKind::Nil => write!(dest, "nil")?,
+    }
+
+    Ok(())
+}
+
+fn write_statement<W: Write>(dest: &mut W, indent: u32, statement: &Statement) -> fmt::Result {
+    for comment in &statement.comments {
+        write!(dest, "// {}\n", comment)?;
+        write_indents(dest, indent)?;
+    }
+
+    match &statement.kind {
+        StatementKind::Assignment {
+            kind,
+            target,
+            value,
+        } => {
+            write_assignable(dest, indent, target)?;
+            write!(
+                dest,
+                " {}= ",
+                match kind {
+                    Op::Nop => "",
+                    Op::Add => "+",
+                    Op::Sub => "-",
+                    Op::Mul => "*",
+                    Op::Div => "/",
+                }
+            )?;
+            write_expression(dest, indent, value)?;
+        }
+        StatementKind::Blob { name, fields } => {
+            write!(dest, "{}", name)?;
+            write!(dest, " :: blob {{\n")?;
+            for (field, ty) in fields {
+                write_indents(dest, indent + 1)?;
+                write!(dest, "{}: ", field)?;
+                write_type(dest, indent, ty)?;
+                write!(dest, ",\n")?;
+            }
+            write_indents(dest, indent)?;
+            write!(dest, "}}")?;
+        }
+        StatementKind::Block { statements } => {
+            write!(dest, "{{\n")?;
+
+            for s in &merge_empty_statements(statements.clone()) {
```

Comment:
Why not make a custom iterator that skips the empty-statements?

Reply:
Interesting...

Reply:
Ah well, the borrows get kinda sad. We'd either need a `Cow` or rewrite so the entire formatter moves instead of borrows. I think the former would be too much work since we _probably_ want the latter anyway.

Reply:
The reason the borrows get sad is because we need to take the comments from the removed statements.